File: unity-mcp-server/src/fileLogger.ts
// ============================================================================
// UnityVision MCP Server - File Logger
// Writes detailed debug logs to a file for troubleshooting
// ============================================================================

import * as fs from 'fs';
import * as path from 'path';

const LOG_FILE = path.join(process.cwd(), 'mcp-debug.log');
const MAX_LOG_SIZE = 5 * 1024 * 1024; // 5MB max

/**
 * Rotate log file if it gets too large
 */
function rotateLogIfNeeded(): void {
  try {
    if (fs.existsSync(LOG_FILE)) {
      const stats = fs.statSync(LOG_FILE);
      if (stats.size > MAX_LOG_SIZE) {
        const backupPath = LOG_FILE + '.old';
        if (fs.existsSync(backupPath)) {
          fs.unlinkSync(backupPath);
        }
        fs.renameSync(LOG_FILE, backupPath);
      }
    }
  } catch {
    // Ignore rotation errors
  }
}

/**
 * Write a log entry to the debug log file
 */
export function fileLog(level: 'DEBUG' | 'INFO' | 'WARN' | 'ERROR', component: string, message: string, data?: unknown): void {
  try {
    rotateLogIfNeeded();
    
    const timestamp = new Date().toISOString();
    let logLine = `[${timestamp}] [${level}] [${component}] ${message}`;
    
    if (data !== undefined) {
      try {
        const dataStr = JSON.stringify(data, null, 0);
        // Truncate very large data
        if (dataStr.length > 2000) {
          logLine += ` | DATA: ${dataStr.substring(0, 2000)}... (truncated, total ${dataStr.length} chars)`;
        } else {
          logLine += ` | DATA: ${dataStr}`;
        }
      } catch {
        logLine += ` | DATA: [unserializable]`;
      }
    }
    
    logLine += '\n';
    
    fs.appendFileSync(LOG_FILE, logLine, 'utf8');
  } catch {
    // Silently fail if we can't write logs
  }
}

/**
 * Log request start
 */
export function logRequestStart(requestId: string, method: string, params: unknown): void {
  fileLog('INFO', 'REQUEST', `START ${requestId} - ${method}`, { params });
}

/**
 * Log request phase
 */
export function logRequestPhase(requestId: string, phase: string, details?: unknown): void {
  fileLog('DEBUG', 'REQUEST', `PHASE ${requestId} - ${phase}`, details);
}

/**
 * Log request end
 */
export function logRequestEnd(requestId: string, method: string, durationMs: number, success: boolean, error?: string): void {
  const level = success ? 'INFO' : 'ERROR';
  fileLog(level, 'REQUEST', `END ${requestId} - ${method} - ${durationMs}ms - ${success ? 'SUCCESS' : 'FAILED'}`, { error });
}

/**
 * Log connection events
 */
export function logConnection(event: string, details?: unknown): void {
  fileLog('INFO', 'CONNECTION', event, details);
}

/**
 * Log errors
 */
export function logError(component: string, message: string, error?: unknown): void {
  fileLog('ERROR', component, message, { error: error instanceof Error ? error.message : error });
}

/**
 * Get log file path
 */
export function getLogFilePath(): string {
  return LOG_FILE;
}

/**
 * Clear the log file
 */
export function clearLog(): void {
  try {
    fs.writeFileSync(LOG_FILE, `=== Log cleared at ${new Date().toISOString()} ===\n`, 'utf8');
  } catch {
    // Ignore
  }
}

// Initialize log on module load
fileLog('INFO', 'INIT', `MCP Debug Logger initialized. Log file: ${LOG_FILE}`);

File: Packages/com.unityvision.bridge/Editor/Bridge/FileLogger.cs
// ============================================================================
// UnityVision MCP Bridge - File Logger
// Writes detailed debug logs to a file for troubleshooting
// ============================================================================

using System;
using System.IO;
using System.Text;
using UnityEngine;

namespace UnityVision.Editor.Bridge
{
    /// <summary>
    /// File-based logger for debugging MCP bridge issues.
    /// Writes to Logs/UnityVision_Debug.log in the project folder.
    /// </summary>
    public static class FileLogger
    {
        private static readonly object _lock = new object();
        private static string _logPath;
        private static bool _initialized;
        private const long MAX_LOG_SIZE = 5 * 1024 * 1024; // 5MB
        
        public static bool Enabled { get; set; } = true;
        
        /// <summary>
        /// Get the log file path
        /// </summary>
        public static string LogPath
        {
            get
            {
                EnsureInitialized();
                return _logPath;
            }
        }
        
        private static void EnsureInitialized()
        {
            if (_initialized) return;
            
            lock (_lock)
            {
                if (_initialized) return;
                
                try
                {
                    string projectPath = Application.dataPath;
                    if (projectPath.EndsWith("/Assets"))
                        projectPath = projectPath.Substring(0, projectPath.Length - 7);
                    
                    string logsDir = Path.Combine(projectPath, "Logs");
                    if (!Directory.Exists(logsDir))
                        Directory.CreateDirectory(logsDir);
                    
                    _logPath = Path.Combine(logsDir, "UnityVision_Debug.log");
                    _initialized = true;
                    
                    // Write header
                    Log("INFO", "INIT", $"FileLogger initialized. Unity {Application.unityVersion}");
                }
                catch (Exception ex)
                {
                    Debug.LogWarning($"[UnityVision] Failed to initialize file logger: {ex.Message}");
                    _initialized = true; // Prevent repeated attempts
                }
            }
        }
        
        /// <summary>
        /// Rotate log file if too large
        /// </summary>
        private static void RotateIfNeeded()
        {
            try
            {
                if (File.Exists(_logPath))
                {
                    var info = new FileInfo(_logPath);
                    if (info.Length > MAX_LOG_SIZE)
                    {
                        string backupPath = _logPath + ".old";
                        if (File.Exists(backupPath))
                            File.Delete(backupPath);
                        File.Move(_logPath, backupPath);
                    }
                }
            }
            catch
            {
                // Ignore rotation errors
            }
        }
        
        /// <summary>
        /// Write a log entry
        /// </summary>
        public static void Log(string level, string component, string message, object data = null)
        {
            if (!Enabled) return;
            
            EnsureInitialized();
            if (string.IsNullOrEmpty(_logPath)) return;
            
            try
            {
                lock (_lock)
                {
                    RotateIfNeeded();
                    
                    var sb = new StringBuilder();
                    sb.Append($"[{DateTime.Now:yyyy-MM-dd HH:mm:ss.fff}] [{level}] [{component}] {message}");
                    
                    if (data != null)
                    {
                        try
                        {
                            string dataStr = JsonUtility.ToJson(data);
                            if (string.IsNullOrEmpty(dataStr) || dataStr == "{}")
                            {
                                dataStr = data.ToString();
                            }
                            if (dataStr.Length > 2000)
                            {
                                sb.Append($" | DATA: {dataStr.Substring(0, 2000)}... (truncated)");
                            }
                            else
                            {
                                sb.Append($" | DATA: {dataStr}");
                            }
                        }
                        catch
                        {
                            sb.Append($" | DATA: {data}");
                        }
                    }
                    
                    sb.AppendLine();
                    File.AppendAllText(_logPath, sb.ToString(), Encoding.UTF8);
                }
            }
            catch
            {
                // Silently fail
            }
        }
        
        /// <summary>
        /// Log request start
        /// </summary>
        public static void LogRequestStart(string requestId, string method)
        {
            Log("INFO", "REQUEST", $"START {requestId} - {method}");
        }
        
        /// <summary>
        /// Log request phase
        /// </summary>
        public static void LogRequestPhase(string requestId, string phase, string details = null)
        {
            Log("DEBUG", "REQUEST", $"PHASE {requestId} - {phase}" + (details != null ? $" - {details}" : ""));
        }
        
        /// <summary>
        /// Log request end
        /// </summary>
        public static void LogRequestEnd(string requestId, string method, int durationMs, bool success, string error = null)
        {
            string level = success ? "INFO" : "ERROR";
            string status = success ? "SUCCESS" : $"FAILED: {error}";
            Log(level, "REQUEST", $"END {requestId} - {method} - {durationMs}ms - {status}");
        }
        
        /// <summary>
        /// Log an error
        /// </summary>
        public static void LogError(string component, string message, Exception ex = null)
        {
            Log("ERROR", component, message + (ex != null ? $" | Exception: {ex.Message}" : ""));
        }
        
        /// <summary>
        /// Clear the log file
        /// </summary>
        public static void Clear()
        {
            EnsureInitialized();
            if (string.IsNullOrEmpty(_logPath)) return;
            
            try
            {
                lock (_lock)
                {
                    File.WriteAllText(_logPath, $"=== Log cleared at {DateTime.Now:O} ===\n", Encoding.UTF8);
                }
            }
            catch
            {
                // Ignore
            }
        }
    }
}
